src/Stratis.Bitcoin.Features.PoA/BasePoAFeatureConsensusRules/PoAHeaderSignatureRule.cs

```diff
@@ -33,6 +53,42 @@ public override void Run(RuleContext context)
 
             if (!this.validator.VerifySignature(pubKey, header))
             {
+                if (this.votingEnabled)
+                {
+                    ChainedHeader currentHeader = context.ValidationContext.ChainedHeaderToValidate;
+
+                    bool mightBeInsufficient = currentHeader.Height - this.chainState.ConsensusTip.Height > this.maxReorg;
+
+                    List<PubKey> modifiedFederation = this.federationManager.GetFederationMembers();
+
+                    foreach (Poll poll in this.votingManager.GetFinishedPolls().Where(x => !x.IsExecuted &&
+                        (x.VotingData.Key == VoteKey.AddFederationMember || x.VotingData.Key == VoteKey.KickFederationMember)))
+                    {
+                        if (currentHeader.Height - poll.PollVotedInFavorBlockData.Height <= this.maxReorg)
+                            // Not applied yet.
+                            continue;
+
+                        var newPubKey = new PubKey(poll.VotingData.Data);
+
+                        if (poll.VotingData.Key == VoteKey.AddFederationMember)
+                            modifiedFederation.Add(newPubKey);
+                        else if (poll.VotingData.Key == VoteKey.KickFederationMember)
+                            modifiedFederation.Remove(newPubKey);
+                    }
+
+                    pubKey = this.slotsManager.GetPubKeyForTimestamp(header.Time, modifiedFederation);
+
+                    if (this.validator.VerifySignature(pubKey, header))
+                    {
+                        this.Logger.LogDebug("Signature verified using updated federation.");
+                        return;
+                    }
+                    else if (mightBeInsufficient)
+                    {
+                        context.ValidationContext.InsufficientHeaderInformation = true;
```

Comment:
This definitely requires more information how InsufficientHeaderInformation is used in the context of POA

Reply:
Added comments